File: src/main/kotlin/com/sifsstudio/botjs/env/api/Bot.java
package com.sifsstudio.botjs.env.api;

import com.sifsstudio.botjs.env.ability.Ability;
import dev.latvian.mods.rhino.ScriptableObject;
import dev.latvian.mods.rhino.annotations.JSFunction;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

public class Bot extends ScriptableObject {
    private Map<String, Ability> nameToAbility = new HashMap<>();

    private Map<Object, Object> memories = new HashMap<>();

    public Bot(Set<Ability> abilities) {
        abilities.forEach(it -> nameToAbility.put(it.getId(), it));
    }

    @Override
    public String getClassName() {
        return "Bot";
    }

    @JSFunction
    public Ability getAbility(String id) {
        return nameToAbility.get(id);
    }

    @JSFunction
    public Object getMemory(Object memId) {
        return memories.get(memId);
    }

    @JSFunction
    public void setMemory(Object memId, Object mem) {
        memories.put(memId, mem);
    }
}

File: src/main/kotlin/com/sifsstudio/botjs/env/api/Bot.kt
package com.sifsstudio.botjs.env.api

import com.sifsstudio.botjs.env.BotEnv
import com.sifsstudio.botjs.env.SuspensionContext
import com.sifsstudio.botjs.env.api.ability.AbilityBase
import com.sifsstudio.botjs.env.block
import com.sifsstudio.botjs.env.suspensionContext
import com.sifsstudio.botjs.env.task.TaskFuture
import java.util.*

class Bot(private val environment: BotEnv, abilities: Map<String, AbilityBase>) {

    @Suppress("unused")
    @JvmField
    val abilities: Map<String, AbilityBase> = Collections.unmodifiableMap(abilities)

    //TODO: see why delegate do not compile

    @Suppress("unused")
    val x
        get() = environment.entity.x

    @Suppress("unused")
    val y
        get() = environment.entity.y

    @Suppress("unused")
    val z
        get() = environment.entity.z

    @Suppress("unused")
    fun <T : Any> block(future: TaskFuture<T>): T = SuspensionContext.invokeSuspend {
        environment.block(future, suspensionContext!!)
    }
}